word_finder.cpp

```diff
@@ -0,0 +1,208 @@
+#include "word_finder.h"
+
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+using namespace std;
+
+Word::Word(unsigned int start, unsigned int end)
+: start(min(start, end))
+, end(max(start, end)) {
+}
+
+CharNode::CharNode(wchar_t c)
+: _c(c)
+, _isEndNode(false)
+, _next() {
+}
+
+CharNode::~CharNode() {
+    _next.clear();
+}
+
+shared_ptr<CharNode> CharNode::nextNode(wchar_t c) const {
+    if (isLeaf()) {
+        return shared_ptr<CharNode>(nullptr);
+    }
+
+    auto it = _next.find(c);
+    return it != _next.end() ? it->second : shared_ptr<CharNode>(nullptr);
+}
+
+bool CharNode::contains(wchar_t c) const {
+    if (isLeaf()) {
+        return false;
+    }
+
+    return _next.find(c) != _next.end();
+}
+
+bool CharNode::isLeaf() const {
+    return _next.empty();
+}
+
+shared_ptr<CharNode> CharNode::add(wchar_t c) {
+    auto it = _next.find(c);
+    if (it != _next.end()) {
+        return it->second;
+    }
+
+    auto newNode = make_shared<CharNode>(c);
+    _next.insert(pair<wchar_t, shared_ptr<CharNode>>(c, newNode));
+    return newNode;
+}
+
+WordFinder::WordFinder()
+: _root(make_shared<CharNode>(L'\0')) {
+}
+
+WordFinder::WordFinder(const list<wstring> &strlist)
+: WordFinder() {
+    for(const wstring &str : strlist) {
+        add(str);
+    }
+}
+
+void WordFinder::add(const wstring &str) {
+    wstring addStr = StringUtils::replace(str, L" ", L"");
+    addStr = StringUtils::trim(addStr);
+    addStr = StringUtils::toLowerCase(addStr);
+
+    auto tmpNode = _root;
+
+    for(unsigned int i = 0U ; i < addStr.length() ; ++i) {
+        wchar_t c = addStr.at(i);
+        auto node = tmpNode->add(c);
+        if (!node) {
+            continue;
+        }
+
+        tmpNode = node;
+        if (i == addStr.length() - 1U) {
+            tmpNode->markEndNode();
```

Comment:
Can be moved after the `for` loop to erase condition checking.
Also, after moving it out of the loop, `i` is only used to get the character, then `for (auto c: addStr)` could be used.